src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/dds/fast_dds.cc

```diff
@@ -0,0 +1,363 @@
+  wqos.reliability().kind = RELIABLE_RELIABILITY_QOS;
+  wqos.publish_mode().kind = ASYNCHRONOUS_PUBLISH_MODE;
+  writer_ = publisher_->create_datawriter(topic_, wqos, listener_.get());
+}
+
+void FastDDSPublisher::InitParticipant() {
+  DomainParticipantQos pqos;
+  pqos.wire_protocol().builtin.discovery_config
+                      .discoveryProtocol = DiscoveryProtocol_t::SIMPLE;
+  pqos.wire_protocol().builtin.discovery_config
+                      .use_SIMPLE_EndpointDiscoveryProtocol = true;
+  pqos.wire_protocol().builtin.discovery_config.m_simpleEDP
+                      .use_PublicationReaderANDSubscriptionWriter = true;
+  pqos.wire_protocol().builtin.discovery_config.m_simpleEDP
+                      .use_PublicationWriterANDSubscriptionReader = true;
+  pqos.wire_protocol().builtin.discovery_config.leaseDuration =
+                       eprosima::fastrtps::c_TimeInfinite;
+  pqos.transport().use_builtin_transports = false;
+  pqos.name("Participant pub" + topic_name_);
+
+  auto transport_descriptor = GetTransportDescriptor(dds_transfer_type_);
+  if (nullptr == transport_descriptor) {
+    LAVA_LOG_ERR("Fatal: Create Transport Fault, exit\n");
+    exit(-1);
+  }
+  pqos.transport().user_transports.push_back(transport_descriptor);
+
+  if (dds_transfer_type_ == DDSTransportType::DDSTCPv4) {
+    Locator_t initial_peer_locator;
+    initial_peer_locator.kind = LOCATOR_KIND_TCPv4;
+    IPLocator::setIPv4(initial_peer_locator, TCPv4_IP);
+    initial_peer_locator.port = TCP_PORT;
+    pqos.wire_protocol().builtin.initialPeersList
+                        .push_back(initial_peer_locator);
+  }
+
+  participant_ = DomainParticipantFactory::get_instance()
+                                           ->create_participant(0, pqos);
+}
+
+bool FastDDSPublisher::Publish(MetaDataPtr metadata) {
+  if (listener_->first_connected_ || listener_->matched_ > 0) {
+    LAVA_DEBUG(LOG_DDS, "FastDDS publisher start publishing...\n");
+    dds_metadata_->nd(metadata->nd);
+    dds_metadata_->type(metadata->type);
+    dds_metadata_->elsize(metadata->elsize);
+    dds_metadata_->total_size(metadata->total_size);
+    memcpy(&dds_metadata_->dims()[0], metadata->dims, sizeof(metadata->dims));
+    memcpy(&dds_metadata_->strides()[0],
+           metadata->strides,
+           sizeof(metadata->strides));
+    dds_metadata_->mdata(std::vector<char>(
+                   reinterpret_cast<char*>(metadata->mdata),
+                   reinterpret_cast<char*>(metadata->mdata) + nbytes_));
+    LAVA_DEBUG(LOG_DDS, "FastDDS publisher copied\n");
+
+    if (writer_->write(dds_metadata_.get()) != ReturnCode_t::RETCODE_OK) {
+      LAVA_LOG_WARN(LOG_DDS, "Publisher write return not OK, Why work?\n");
+    } else {
+      LAVA_DEBUG(LOG_DDS, "Publish a data\n");
+    }
+    return true;
+  }
+  return false;
+}
+
+void FastDDSPublisher::Stop() {
+  LAVA_LOG(LOG_DDS, "Stop FastDDS Publisher, waiting unmatched...\n");
+  while (listener_->matched_ > 0) {
```

Comment:
if others are ROS2 node, must to wait matched node to zero at FastDDSPublisher?

Reply:
After publisher exist, the subscriber cannot recv the data. I can check the qos config for publisher stop and subscriber still recv...